File: iosApp/iosApp/Presentation/Main/CafeManagement/CastList/CastListViewModel.swift
//
//  CastListViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/07/05.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class CastListViewModel: ObservableObject {
    private let cafeId: String

    private let getCafeCastPageUseCase: GetCafeCastPageUseCase

    private let deleteCastUseCase: DeleteCastUseCase

    private let castEventPublisher: CastEventPublisher

    @Published private(set) var uiState = CastListUiState()

    let event = PassthroughSubject<CastListEvent, Never>()

    private var observeCastEventTask: Task<Void, Never>?

    private func loadCastPage(cursor: String?, append: Bool) {
        Task {
            uiState.isLoading = !append && uiState.casts.isEmpty
            uiState.isLoadingMore = append
            if !append {
                uiState.infoMessage = nil
            }

            do {
                let result = try await getCafeCastPageUseCase.invoke(
                    cafeId: cafeId,
                    cursor: cursor,
                    pageSize: getCafeCastPageUseCase.defaultPageSize()
                )

                if let success = result as? AppResultSuccess<AnyObject>,
                   let page = success.data as? PagedResult<CafeCastPreview> {
                    let items = (page.items as? [CafeCastPreview]) ?? []
                    uiState.casts = append ? (uiState.casts + items) : items
                    uiState.nextCursor = page.nextCursor
                    uiState.hasMoreCasts = page.hasNext
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                } else {
                    uiState.isLoading = false
                    uiState.isLoadingMore = false
                    uiState.infoMessage = "castlist_info_load_failed"
                }
            } catch {
                uiState.isLoading = false
                uiState.isLoadingMore = false
                uiState.infoMessage = "castlist_info_load_failed"
            }
        }
    }

    private func refreshCasts() {
        loadCastPage(cursor: nil, append: false)
    }

    private func clickLoadMoreCasts() {
        guard uiState.hasMoreCasts, !uiState.isLoadingMore else { return }
        loadCastPage(cursor: uiState.nextCursor, append: true)
    }

    private func clickBack() {
        event.send(.navigateBack)
    }

    private func clickAddCast() {
        event.send(.navigateToCastEdit(cafeId: cafeId, castId: nil))
    }

    private func changeSearchQuery(_ value: String) {
        uiState.searchQuery = value
    }

    private func clickCast(_ castId: String) {
        event.send(.navigateToCastEdit(cafeId: cafeId, castId: castId))
    }

    private func clickCastSchedule(_ castId: String) {
        event.send(.navigateToSchedule(castId: castId))
    }

    private func clickDeleteCast(_ castId: String) {
        uiState.deleteTargetCastId = castId
        uiState.infoMessage = nil
    }

    private func dismissDeleteCastDialog() {
        uiState.deleteTargetCastId = nil
    }

    private func confirmDeleteCast() {
        guard let deleteTargetCastId = uiState.deleteTargetCastId else {
            uiState.deleteTargetCastId = nil
            return
        }

        Task {
            do {
                let result = try await deleteCastUseCase.invoke(castId: deleteTargetCastId)

                if result is AppResultSuccess<AnyObject> {
                    uiState.deleteTargetCastId = nil
                    uiState.infoMessage = "castlist_info_cast_deleted"
                } else if let failure = result as? AppResultFailure {
                    uiState.deleteTargetCastId = nil
                    uiState.infoMessage = "\(failure.error)"
                }
            } catch {
                uiState.deleteTargetCastId = nil
                uiState.infoMessage = error.localizedDescription
            }
        }
    }

    private func dismissInfoMessage() {
        uiState.infoMessage = nil
    }

    private func observeCastEvent() {
        observeCastEventTask?.cancel()
        observeCastEventTask = Task {
            do {
                for try await event in asyncSequence(for: castEventPublisher.events) {
                    switch event {
                    case let event as Shared.CastEvent.Created:
                        if event.cafeId == self.cafeId {
                            self.refreshCasts()
                        }
                    case let event as Shared.CastEvent.Updated:
                        if event.cafeId == self.cafeId {
                            self.uiState.casts = self.uiState.casts.map { preview in
                                guard preview.id == event.cast.id else { return preview }
                                return CafeCastPreview(
                                    id: preview.id,
                                    name: event.cast.name,
                                    isOnShift: preview.isOnShift,
                                    profileImage: event.cast.profileImage
                                )
                            }
                        }
                    case let event as Shared.CastEvent.Deleted:
                        if event.cafeId == self.cafeId {
                            self.uiState.casts.removeAll { $0.id == event.castId }
                            if self.uiState.deleteTargetCastId == event.castId {
                                self.uiState.deleteTargetCastId = nil
                            }
                        }
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    func onAction(_ action: CastListAction) {
        switch action {
        case .clickBack:
            clickBack()
        case .clickAddCast:
            clickAddCast()
        case .changeSearchQuery(let value):
            changeSearchQuery(value)
        case .clickCast(let castId):
            clickCast(castId)
        case .clickCastSchedule(let castId):
            clickCastSchedule(castId)
        case .clickDeleteCast(let castId):
            clickDeleteCast(castId)
        case .confirmDeleteCast:
            confirmDeleteCast()
        case .dismissDeleteCastDialog:
            dismissDeleteCastDialog()
        case .clickLoadMoreCasts:
            clickLoadMoreCasts()
        case .dismissInfoMessage:
            dismissInfoMessage()
        }
    }

    init(
        cafeId: String,
        getCafeCastPageUseCase: GetCafeCastPageUseCase = KoinInitializerKt.resolveGetCafeCastPageUseCase(),
        deleteCastUseCase: DeleteCastUseCase = KoinInitializerKt.resolveDeleteCastUseCase(),
        castEventPublisher: CastEventPublisher = KoinInitializerKt.resolveCastEventPublisher()
    ) {
        self.cafeId = cafeId
        self.getCafeCastPageUseCase = getCafeCastPageUseCase
        self.deleteCastUseCase = deleteCastUseCase
        self.castEventPublisher = castEventPublisher

        observeCastEvent()
        refreshCasts()
    }

    deinit {
        observeCastEventTask?.cancel()
        observeCastEventTask = nil
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/castlist/CastListViewModel.kt
package com.hhp227.concafe.presentation.main.cafemanagement.castlist

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.event.publisher.CastEventPublisher
import com.hhp227.concafe.domain.event.CastEvent as CastDomainEvent
import com.hhp227.concafe.domain.usecase.DeleteCastUseCase
import com.hhp227.concafe.domain.usecase.GetCafeCastPageUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class CastListViewModel(
    private val cafeId: String,
    private val getCafeCastPageUseCase: GetCafeCastPageUseCase,
    private val deleteCastUseCase: DeleteCastUseCase,
    private val castEventPublisher: CastEventPublisher
) : ViewModel() {
    private val _uiState = MutableStateFlow(CastListUiState())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<CastListEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private var observeCastEventJob: Job? = null

    private fun loadCastPage(cursor: String?, append: Boolean) {
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = !append && it.casts.isEmpty(),
                    isLoadingMore = append,
                    infoMessage = if (append) it.infoMessage else null
                )
            }
            when (val result = getCafeCastPageUseCase.invoke(cafeId, cursor, getCafeCastPageUseCase.defaultPageSize())) {
                is AppResult.Success -> {
                    _uiState.update { state ->
                        val mergedItems = if (append) state.casts + result.data.items else result.data.items

                        state.copy(
                            casts = mergedItems,
                            nextCursor = result.data.nextCursor,
                            hasMoreCasts = result.data.hasNext,
                            isLoading = false,
                            isLoadingMore = false
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            isLoadingMore = false,
                            infoMessage = "castlist_info_load_failed"
                        )
                    }
                }
            }
        }
    }

    private fun refreshCasts() {
        loadCastPage(cursor = null, append = false)
    }

    private fun clickLoadMoreCasts() {
        val currentState = _uiState.value

        if (currentState.isLoadingMore || !currentState.hasMoreCasts) {
            return
        }
        loadCastPage(cursor = currentState.nextCursor, append = true)
    }

    private fun clickBack() {
        viewModelScope.launch {
            _event.emit(CastListEvent.NavigateBack)
        }
    }

    private fun clickAddCast() {
        viewModelScope.launch {
            _event.emit(CastListEvent.NavigateToCastEdit(cafeId = cafeId))
        }
    }

    private fun changeSearchQuery(value: String) {
        _uiState.update { it.copy(searchQuery = value) }
    }

    private fun clickCast(castId: String) {
        viewModelScope.launch {
            _event.emit(CastListEvent.NavigateToCastEdit(cafeId = cafeId, castId = castId))
        }
    }

    private fun clickCastSchedule(castId: String) {
        viewModelScope.launch {
            _event.emit(CastListEvent.NavigateToSchedule(castId))
        }
    }

    private fun clickDeleteCast(castId: String) {
        _uiState.update { it.copy(deleteTargetCastId = castId, infoMessage = null) }
    }

    private fun dismissDeleteCastDialog() {
        _uiState.update { it.copy(deleteTargetCastId = null) }
    }

    private fun confirmDeleteCast() {
        val deleteTargetCastId = _uiState.value.deleteTargetCastId

        if (deleteTargetCastId == null) {
            _uiState.update { it.copy(deleteTargetCastId = null) }
        } else {
            viewModelScope.launch {
                when (val result = deleteCastUseCase.invoke(deleteTargetCastId)) {
                    is AppResult.Success -> {
                        _uiState.update {
                            it.copy(
                                deleteTargetCastId = null,
                                infoMessage = "castlist_info_cast_deleted"
                            )
                        }
                    }
                    is AppResult.Failure -> {
                        _uiState.update {
                            it.copy(
                                deleteTargetCastId = null,
                                infoMessage = result.error.toString()
                            )
                        }
                    }
                }
            }
        }
    }

    private fun dismissInfoMessage() {
        _uiState.update { it.copy(infoMessage = null) }
    }

    private fun observeCastEvent() {
        observeCastEventJob?.cancel()
        observeCastEventJob = viewModelScope.launch {
            castEventPublisher.events.collectLatest { event ->
                when (event) {
                    is CastDomainEvent.Created -> if (event.cafeId == cafeId) {
                        refreshCasts()
                    }
                    is CastDomainEvent.Updated -> if (event.cafeId == cafeId) {
                        _uiState.update { state ->
                            state.copy(
                                casts = state.casts.map { preview ->
                                    if (preview.id == event.cast.id) {
                                        preview.copy(
                                            name = event.cast.name,
                                            profileImage = event.cast.profileImage
                                        )
                                    } else {
                                        preview
                                    }
                                }
                            )
                        }
                    }
                    is CastDomainEvent.Deleted -> if (event.cafeId == cafeId) {
                        _uiState.update { state ->
                            state.copy(
                                casts = state.casts.filterNot { it.id == event.castId },
                                deleteTargetCastId = state.deleteTargetCastId?.takeUnless { it == event.castId }
                            )
                        }
                    }
                }
            }
        }
    }

    fun onAction(action: CastListAction) {
        when (action) {
            CastListAction.ClickBack -> clickBack()
            CastListAction.ClickAddCast -> clickAddCast()
            is CastListAction.ChangeSearchQuery -> changeSearchQuery(action.value)
            is CastListAction.ClickCast -> clickCast(action.castId)
            is CastListAction.ClickCastSchedule -> clickCastSchedule(action.castId)
            is CastListAction.ClickDeleteCast -> clickDeleteCast(action.castId)
            CastListAction.ConfirmDeleteCast -> confirmDeleteCast()
            CastListAction.DismissDeleteCastDialog -> dismissDeleteCastDialog()
            CastListAction.ClickLoadMoreCasts -> clickLoadMoreCasts()
            CastListAction.DismissInfoMessage -> dismissInfoMessage()
        }
    }

    init {
        observeCastEvent()
        refreshCasts()
    }

    override fun onCleared() {
        observeCastEventJob?.cancel()
        observeCastEventJob = null
        super.onCleared()
    }
}
